crypto/rsa/rsa_sp800_56b_gen.c

```diff
@@ -0,0 +1,327 @@
+
+    for (;;) {
+        /* (Step 4) Generate p, Xp */
+        if (!bn_rsa_fips186_4_gen_prob_primes(rsa->p, Xpo, p1, p2, Xp, Xp1, Xp2,
+                                              nbits, e, ctx, cb))
+            goto err;
+        for(;;) {
+            /* (Step 5) Generate q, Xq*/
+            if (!bn_rsa_fips186_4_gen_prob_primes(rsa->q, Xqo, q1, q2, Xq, Xq1,
+                                                  Xq2, nbits, e, ctx, cb))
+                goto err;
+
+            /* (Step 6) |Xp - Xq| > 2^(nbitlen/2 - 100) */
+            ok = rsa_check_pminusq_diff(tmp, Xpo, Xqo, nbits);
+            if (ok < 0)
+                goto err;
+            if (ok == 0)
+                continue;
+
+            /* (Step 6) |p - q| > 2^(nbitlen/2 - 100) */
+            ok = rsa_check_pminusq_diff(tmp, rsa->p, rsa->q, nbits);
+            if (ok < 0)
+                goto err;
+            if (ok == 0)
+                continue;
+            goto end; /* successfully finished */
+        }
+    }
+end:
+    ret = 1;
+err:
+    /* Zeroize any internally generated values that are not returned */
+    if (Xpo != Xpout)
+        BN_clear(Xpo);
+    if (Xqo != Xqout)
+        BN_clear(Xqo);
+    BN_clear(tmp);
+
+    BN_CTX_end(ctx);
+    return ret;
+}
+
+/* SP800-56Br1 6.3.1.1 (Steps 1a-1b) */
+int rsa_sp800_56b_validate_strength(int nbits, int strength)
+{
+    if (strength != -1) {
+        if (strength == 112) {
+            if (nbits != 2048) {
+                RSAerr(RSA_F_RSA_SP800_56B_VALIDATE_STRENGTH,
+                       RSA_R_INVALID_MODULUS);
+                return 0;
+            }
+        } else if (strength == 128) {
+            if (nbits != 3072) {
+                RSAerr(RSA_F_RSA_SP800_56B_VALIDATE_STRENGTH,
+                       RSA_R_INVALID_MODULUS);
+                return 0;
+            }
+        } else {
+            RSAerr(RSA_F_RSA_SP800_56B_VALIDATE_STRENGTH,
+                   RSA_R_INVALID_STRENGTH);
+            return 0;
+        }
+    }
+    return 1;
+}
+
+/*
+ * SP800-56Br1
+ * 6.3.1.1 rsakpg1 - basic (Steps 3-4)
+ * 6.3.1.3 rsakpg1 - crt   (Step 5)
+ *
+ * Using p & q, calculate other required parameters such as n, d.
+ * as well as the CRT parameters dP, dQ, qInv.
+ *
+ * Note:
+ *   There is a small chance that the generated d will be too small.
+ * Returns -1 = error,
+ *          0 = d is too small,
+ *          1 = success.
+ */
+int rsa_sp800_56b_derive_params_from_pq(RSA *rsa, int nbits,
+                                        const BIGNUM *e, BN_CTX *ctx)
+{
+    int ret = -1;
+    BN_CTX *ctx2 = NULL;
+    BIGNUM *p1, *q1, *lcm, *p1q1, *gcd;
+
+    BN_CTX_start(ctx);
+    p1 = BN_CTX_get(ctx);
+    q1 = BN_CTX_get(ctx);
+    lcm = BN_CTX_get(ctx);
+    p1q1 = BN_CTX_get(ctx);
+    gcd = BN_CTX_get(ctx);
+    if (gcd == NULL)
+        goto err;
+
+    /* LCM((p-1, q-1)) */
+    if (rsa_get_lcm(ctx, rsa->p, rsa->q, lcm, gcd, p1, q1, p1q1) != 1)
+        goto err;
+
+    ctx2 = BN_CTX_new();
```

Comment:
Why do we need this?

Reply:
I will try to remove it - this code was block copied from the original code.